### PR TITLE
Combine Reader/Feeder together in trainer.train

### PR DESCRIPTION
**Pull request by reyoung**
Depends on #1476

### CODE REVIEW

File: demo/mnist/api_train_v2.py [hunk 1]
```diff
@@ -45,17 +36,24 @@ def event_handler(event):
...
-                  )
+    reader = paddle.reader.batched(
+        paddle.reader.shuffle(
+            paddle.dataset.mnist.train_creator(), buf_size=8192),
+        batch_size=32)
+
+    trainer.train(
+        train_reader=paddle.reader.batched(
+            paddle.reader.shuffle(paddle.dataset.mnist.train_creator(),
+                                  buf_size=8192), batch_size=32),
```
Comment (qingqing01) [inline demo/mnist/api_train_v2.py]: train_reader=...    ->  train_reader=reader

File: python/paddle/v2/trainer.py [hunk 2]
```diff
@@ -62,27 +62,23 @@ def __init__(self, update_equation):
         self.__optimizer__ = update_equation
 
     def train(self,
-              train_data_reader,
+              train_reader,
```
Comment (qingqing01) [inline python/paddle/v2/trainer.py]: 和ITrainer里的名字不一样~

File: demo/mnist/api_train_v2.py (hunk 1 above)
Comment (reyoung) [inline demo/mnist/api_train_v2.py]: Done.

File: python/paddle/v2/trainer.py (hunk 2 above)
Comment (reyoung) [inline python/paddle/v2/trainer.py]: Done.

File: python/setup.py.in [hunk 3]
```diff
@@ -5,13 +5,19 @@ packages=['paddle',
...
 setup(name='paddle',
       version='${PADDLE_VERSION}',
       description='Parallel Distributed Deep Learning',
       packages=packages,
       package_dir={
           '': '${CMAKE_CURRENT_SOURCE_DIR}'
-      }
+      },
+      install_requires = [
+          'scikit-learn>=0.18.0',
```
Comment (helinwang) [inline python/setup.py.in]: 什么模块需要这两个package？我记得数据模块需要的是`nltk`不是这两个。这个commit看不出来哪里需要scikit和scipy。

File: python/setup.py.in [hunk 4]
```diff
@@ -5,13 +5,19 @@ packages=['paddle',
...
+          'scipy>=0.18.0',
```
Comment (helinwang) [inline python/setup.py.in]: 什么模块需要`scipy`？这个commit看不出来哪里需要scipy。我搜了下paddle也没有哪个地方用到了`scipy`

**Comment by helinwang** [comment]
test没通过，在这里：https://travis-ci.org/PaddlePaddle/Paddle/jobs/205778169#L2543
> test test_data_feeder.py
Traceback (most recent call last):
  File "test_data_feeder.py", line 20, in <module>
    from paddle.v2 import data_type
  File "/home/travis/build/PaddlePaddle/Paddle/python/paddle/v2/__init__.py", line 22, in <module>
    from . import dataset
  File "/home/travis/build/PaddlePaddle/Paddle/python/paddle/v2/dataset/__init__.py", line 1, in <module>
    import mnist
  File "/home/travis/build/PaddlePaddle/Paddle/python/paddle/v2/dataset/mnist.py", line 21, in <module>
    "MNIST original", data_home=DATA_HOME)
  File "/opt/python/2.7.12/lib/python2.7/site-packages/sklearn/datasets/mldata.py", line 158, in fetch_mldata
    matlab_dict = io.loadmat(matlab_file, struct_as_record=True)
  File "/opt/python/2.7.12/lib/python2.7/site-packages/scipy/io/matlab/mio.py", line 136, in loadmat
    matfile_dict = MR.get_variables(variable_names)
  File "/opt/python/2.7.12/lib/python2.7/site-packages/scipy/io/matlab/mio5.py", line 292, in get_variables
    res = self.read_var_array(hdr, process)
  File "/opt/python/2.7.12/lib/python2.7/site-packages/scipy/io/matlab/mio5.py", line 252, in read_var_array
    return self._matrix_reader.array_from_header(header, process)
  File "scipy/io/matlab/mio5_utils.pyx", line 673, in scipy.io.matlab.mio5_utils.VarReader5.array_from_header (scipy/io/matlab/mio5_utils.c:7119)
  File "scipy/io/matlab/mio5_utils.pyx", line 703, in scipy.io.matlab.mio5_utils.VarReader5.array_from_header (scipy/io/matlab/mio5_utils.c:6244)
  File "scipy/io/matlab/mio5_utils.pyx", line 776, in scipy.io.matlab.mio5_utils.VarReader5.read_real_complex (scipy/io/matlab/mio5_utils.c:7572)
  File "scipy/io/matlab/mio5_utils.pyx", line 448, in scipy.io.matlab.mio5_utils.VarReader5.read_numeric (scipy/io/matlab/mio5_utils.c:4323)
  File "scipy/io/matlab/mio5_utils.pyx", line 353, in scipy.io.matlab.mio5_utils.VarReader5.read_element (scipy/io/matlab/mio5_utils.c:3913)
  File "scipy/io/matlab/streams.pyx", line 327, in scipy.io.matlab.streams.FileStream.read_string (scipy/io/matlab/streams.c:4992)
IOError: could not read bytes

File: python/setup.py.in (hunk 4 above)
Comment (reyoung) [inline python/setup.py.in]: scikit-learn 依赖了scipy。特别是在dataset那块的代码用到了scipy

Comment (reyoung) [inline python/setup.py.in]: 这个修改会随着 #1476 的合入，而去掉。

我等#1476合入之后，再修改一下吧。

File: python/paddle/v2/dataset/mnist.py [hunk 5]
```diff
@@ -1,39 +1,64 @@
...
-    return __mnist_reader_creator__(X_train, y_train)
+            for i in xrange(buffer_size):
+                yield images[i, :], labels[i]
 
+        m.terminate()
+        l.terminate()
 
-def test_creator():
-    return __mnist_reader_creator__(X_test, y_test)
+    return reader()
```
Comment (helinwang) [inline python/paddle/v2/dataset/mnist.py]: Maybe return `reader` instead of `reader()`?

File: python/paddle/v2/dataset/mnist.py [hunk 6]
```diff
@@ -1,39 +1,72 @@
...
+TRAIN_LABEL_MD5 = 'd53e105ee54ea40749a09fcbcd1e9432'
+
+
+def reader_creator(image_filename, label_filename, buffer_size):
     def reader():
-        n_samples = data.shape[0]
-        for i in xrange(n_samples):
-            yield (data[i] / 255.0).astype(numpy.float32), int(target[i])
+        if platform.system() == 'Darwin':
+            zcat_cmd = 'gzcat'
```
Comment (reyoung) [inline python/paddle/v2/dataset/mnist.py]: @wangkuiyi 

MacOS和Linux下的zcat命令不一样。。稍微改了一下。

File: python/paddle/v2/dataset/mnist.py [hunk 7]
```diff
@@ -20,12 +20,19 @@
 
 def reader_creator(image_filename, label_filename, buffer_size):
     def reader():
+        if platform.system() == 'Darwin':
```
Comment (wangkuiyi) [inline python/paddle/v2/dataset/mnist.py]: 需要这样吗？我的mac里gzcat和zcat都有呢？

File: python/paddle/v2/reader/decorator.py [hunk 8]
```diff
@@ -191,3 +191,25 @@ def data_reader():
...
+
+def batched(reader, batch_size):
+    """
+    Create a batched reader.
+    :param reader: the data reader to read from.
+    :param batch_size: batch_size
+    :return: the batched reader.
+    """
+
+    def __impl__():
```
Comment (wangkuiyi) [inline python/paddle/v2/reader/decorator.py]: __impl__ ==> batched_reader 
保持code style。

File: python/paddle/v2/dataset/mnist.py (hunk 7 above)
Comment (reyoung) [inline python/paddle/v2/dataset/mnist.py]: 我的mac里面确实有zcat。但是和gzcat的行为不一样。导致没有办法解压。

File: python/paddle/v2/reader/decorator.py (hunk 8 above)
Comment (reyoung) [inline python/paddle/v2/reader/decorator.py]: Done.

**Pull request merged**